### ISSUE TITLE
Herbstluftwm Help

### ISSUE THREAD

**Issue by rjahrj** [nix-community/home-manager#2884]
Hello, I currently have a working herbstluftwm environment. It is installed using `configuration.nix`. I didn't really notice problems until I tried getting dmenu working. Pkgs I installed thru home-manager wouldn't show up on dmenu. So I tried looking more into it, and I think it has to do with my WM not being configured with home-manager. If I can get some help setting that up, I would definitely appreciate it

**Comment by berbiche** [comment]
Hi @rjahrj,

First, can you confirm what Linux distribution you use?

Can you also confirm that `$HOME/.nix-profile/bin` is in your `$PATH`?
If not, can you share the content of this variable.

**Comment by rjahrj** [comment]
@berbiche hello
thanks for reaching out. I am using NixOS 20.11 and yes I can confirm that `.nix-profile/bin` is in my path

**Comment by rjahrj** [comment]
This is the error I get when I now try to use `home-manager switch`

`error : list index 0 is out of bounds, at /home/Rj/.nix-defexpr/channels/home-manager/modules/services/window-managers/herbstluftwm.nix`

![20220413_075847](https://user-images.githubusercontent.com/20189326/163178827-32c26c9d-5ad6-494f-821b-e3fb9d8193ba.jpg)
![20220413_075758](https://user-images.githubusercontent.com/20189326/163178845-e15c02a3-7d48-48e8-837f-1c043fa9917a.jpg)

**Comment by berbiche** [comment]
Sorry for the late reply.

You have to set at least one tag in the configuration with `xsession.windowManager.herbstluftwm.tags = [ "something" ];`.

I think the issue is due to this line in Home Manager:
https://github.com/nix-community/home-manager/blob/8ab155c61f5821ffda723de88b0009769771d4f2/modules/services/window-managers/herbstluftwm.nix#L148-L152

cc: @olmokramer

**Comment by rjahrj** [comment]
@berbiche thanks that definitely allowed me to enable Herbst on HM, now the problem is the autostart file. Or the Nix equivalent? 
![Screenshot_20220417-002800](https://user-images.githubusercontent.com/20189326/163700550-ed486fed-6d8c-4867-be1d-32be8e84dfad.png)
This just gives me the red border and nothing works 🤦🏽

**Comment by berbiche** [comment]
Sorry, I'm not sure. I have no experience with Herbstluftwm.

You can change your configuration iteratively and look at the generated configuration file to fix anything until you have a 1-to-1 mapping of your shell script with the nix expression.

**Comment by olmokramer** [comment]
Sorry for the late reply, was on vacation for a couple of days. Could you share the autostart script you're trying to recreate @rjahrj ? What do you mean with "nothing works"?

**Comment by rjahrj** [comment]
@olmokramer perfectly fine, actually I am using the default config from the Herbstluftwm git repo. But I was at one point trying to load it using the `xdg.configFile` option while keeping it in my dotfiles. However both files become read-only 🤦🏽. So what is the default way the define and manage the autostart file using Home-manager? Please and thank you. At this point I just want to know lol 🤦🏽

**Comment by olmokramer** [comment]
If you don't want to manage your Herbstluftwm config with Home Manager I would recommend setting

```
xsession.windowManager.herbstluftwm.enable = false
```

and make sure that Herbstluft is only started after Home Manger has done its setup of the `PATH` environment variable, for example by setting

```
xsession.windowManager.command = "${pkgs.herbstluftwm}/bin/herbstluftwm --locked";
```

Otherwise, if you do want to manage your Herbstluftwm config with Home Manager, you will need to set the settings and keybindings in your Home Manager configuration.

**Comment by rjahrj** [comment]
Yes I want to use HM, so pretty much I need to convert the autostart file
into the separate HM Herbstluftwm functions?

**Comment by rjahrj** [comment]
@olmokramer thanks that definitely helps a lot
Where would I put stuff like `hc detect_monitors` I tried putting it in `herbstluftwm.extraConfig` but that didn't work 😞

**Comment by rjahrj** [comment]
@olmokramer sorry for spamming you with questions 😭 but what about tag switching?
![Screenshot_20220421-185448](https://user-images.githubusercontent.com/20189326/164564503-31db6353-4ca1-48a7-b773-01380a92867b.png)

**Comment by berbiche** [comment]
@rjahrj You can use the `keybinds` option and the `tags` option.

I recommend looking at the documentation for the options here: https://nix-community.github.io/home-manager/options.html#opt-xsession.windowManager.herbstluftwm.enable

The shell code can be translated to:

``` nix
let
  # A list is used because it preserves ordering
  tags = [
    { name = "tag 1"; key = "something"; }
    { name = "tag 2"; }
    ...
  ];
  # Attribute set of { <keybind> = "use <tag>"; }
  keybinds = let
    tagsWithKeybinds = lib.filter (lib.hasAttr "key") tags;
  in lib.listToAttrs (map (x: lib.nameValuePair x.key "use ${x.name}") tagsWithKeybinds);
in
{
  xsession.windowManager.herbstluftwm.keybinds = keybinds;
  xsession.windowManager.herbstluftwm.tags = lib.catAttrs "name" tags;
}
```

I also recommend reading the nixpkgs manual and getting acquainted with the library functions: https://nixos.org/manual/nixpkgs/unstable/#chap-functions

**Comment by olmokramer** [comment]
@rjahrj Sorry for the late reply, work was insanely busy the past couple of days. Yes you should put that in the `herbstluftwm.extraConfig` section, but you should use `herbstclient` instead of `hc`. If you want to use `hc` you can put `alias hc=herbstclient` at the top of `herbstluftwm.extraConfig`. For the tag switching keybinds you'd use something like @berbiche wrote.

**Comment by Qwerty-Space** [comment]
> @rjahrj You can use the `keybinds` option and the `tags` option.
> 
> I recommend looking at the documentation for the options here: https://nix-community.github.io/home-manager/options.html#opt-xsession.windowManager.herbstluftwm.enable
> 
> The shell code can be translated to:
> 
> ```nix
> let
>   # A list is used because it preserves ordering
>   tags = [
>     { name = "tag 1"; key = "something"; }
>     { name = "tag 2"; }
>     ...
>   ];
>   # Attribute set of { <keybind> = "use <tag>"; }
>   keybinds = let
>     tagsWithKeybinds = lib.filter (lib.hasAttr "key") tags;
>   in lib.listToAttrs (map (x: lib.nameValuePair x.key "use ${x.name}") tagsWithKeybinds);
> in
> {
>   xsession.windowManager.herbstluftwm.keybinds = keybinds;
>   xsession.windowManager.herbstluftwm.tags = lib.catAttrs "name" tags;
> }
> ```
> 
> I also recommend reading the nixpkgs manual and getting acquainted with the library functions: https://nixos.org/manual/nixpkgs/unstable/#chap-functions

Can that keybinds variable be used in a list of other keybindings?

For example
``` Mod-bracketright = "cycle +1";
Mod-bracketleft = "cycle -1";
Mod-i = "jumpto urgent";

keybinds;
```
